File: src/main/java/io/openfuture/openmessanger/security/CognitoAuthenticationProvider.java
package io.openfuture.openmessanger.security;

import java.math.BigInteger;
import java.net.URL;
import java.security.Key;
import java.security.KeyFactory;
import java.security.spec.RSAPublicKeySpec;
import java.util.Base64;
import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.stereotype.Component;

import com.fasterxml.jackson.databind.ObjectMapper;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jws;
import io.jsonwebtoken.JwtParser;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.LocatorAdapter;
import io.jsonwebtoken.ProtectedHeader;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Component
public class CognitoAuthenticationProvider implements AuthenticationProvider {

    @Value("${jwks.url}")
    private String jwksUrl;

    @SneakyThrows
    @Override
    public Authentication authenticate(final Authentication authentication) throws AuthenticationException {
        String token = (String) authentication.getCredentials();

        final JwtParser parser = Jwts.parser().keyLocator(new MyKeyLocator(jwksUrl)).build();
        final Jws<Claims> claimsJws;
        try {
            claimsJws = parser.parseSignedClaims(token);
        } catch (Exception e) {
            throw new BadCredentialsException("Unable to parse access token", e);
        }

        final String username = claimsJws.getPayload().get("username", String.class);
        final String scope = claimsJws.getPayload().get("scope", String.class);
        final SimpleGrantedAuthority simpleGrantedAuthority = new SimpleGrantedAuthority(scope);

        return new CognitoAuthenticationToken(username, token, List.of(simpleGrantedAuthority));
    }

    @Override
    public boolean supports(final Class<?> authentication) {
        return CognitoAuthenticationToken.class.isAssignableFrom(authentication);
    }

    public static class MyKeyLocator extends LocatorAdapter<Key> {

        private final String jwksUrl;

        public MyKeyLocator(final String jwksUrl) {
            this.jwksUrl = jwksUrl;
        }

        @SneakyThrows
        @Override
        protected Key locate(final ProtectedHeader header) {
            final String keyId = header.getKeyId();

            final ObjectMapper objectMapper = new ObjectMapper();
            Map<String, Object> jwksMap = objectMapper.readValue(new URL(jwksUrl), Map.class);

            // Extract the keys array from the JWKS
            List<Map<String, Object>> keys = (List<Map<String, Object>>) jwksMap.get("keys");

            // Find the key with the matching key ID (kid)
            for (Map<String, Object> key : keys) {
                String kid = (String) key.get("kid");
                if (kid != null && kid.equals(keyId)) {
                    // Convert the key to a PublicKey (e.g., RSA)
                    String alg = (String) key.get("alg");
                    String e = (String) key.get("e");
                    String n = (String) key.get("n");
                    byte[] eBytes = Base64.getUrlDecoder().decode(e);
                    byte[] nBytes = Base64.getUrlDecoder().decode(n);

                    BigInteger modulus = new BigInteger(1, nBytes);
                    BigInteger exponent = new BigInteger(1, eBytes);

                    // Construct RSA public key
                    RSAPublicKeySpec keySpec = new RSAPublicKeySpec(modulus, exponent);
                    KeyFactory keyFactory = KeyFactory.getInstance("RSA");
                    return keyFactory.generatePublic(keySpec);
                }
            }
            return null;
        }
    }

}

File: src/main/kotlin/io/openfuture/openmessanger/security/CognitoAuthenticationProvider.kt
package io.openfuture.openmessanger.security

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import io.jsonwebtoken.*
import io.openfuture.openmessanger.security.CognitoAuthenticationToken
import lombok.SneakyThrows
import lombok.extern.slf4j.Slf4j
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.authentication.*
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.stereotype.Component
import org.springframework.util.StringUtils
import java.math.BigInteger
import java.net.URL
import java.security.Key
import java.security.KeyFactory
import java.security.spec.RSAPublicKeySpec
import java.util.*

@Slf4j
@Component
class CognitoAuthenticationProvider : AuthenticationProvider {
    @Value("\${jwks.url}")
    private val jwksUrl: String? = null
    @SneakyThrows
    @Throws(AuthenticationException::class)
    override fun authenticate(authentication: Authentication): Authentication {
        val token = authentication.credentials as String
        if (!StringUtils.hasText(token)) {
            throw BadCredentialsException("Token not provided")
        }
        val parser = Jwts.parser().keyLocator(MyKeyLocator(jwksUrl)).build()
        val claimsJws: Jws<Claims> = try {
            parser.parseSignedClaims(token)
        } catch (e: Exception) {
            throw BadCredentialsException("Unable to parse access token", e)
        }
        val username = claimsJws.payload.get("username", String::class.java)
        val scope = claimsJws.payload.get("scope", String::class.java)
        val simpleGrantedAuthority = SimpleGrantedAuthority(scope)
        return CognitoAuthenticationToken(username, token, listOf(simpleGrantedAuthority))
    }

    override fun supports(authentication: Class<*>?): Boolean {
        return CognitoAuthenticationToken::class.java.isAssignableFrom(authentication)
    }

    class MyKeyLocator(private val jwksUrl: String?) : LocatorAdapter<Key?>() {
        @SneakyThrows
        override fun locate(header: ProtectedHeader): Key? {
            val keyId = header.keyId
            val objectMapper = jacksonObjectMapper()

            val jwksMap: Map<String, Any> = objectMapper.readValue(URL(jwksUrl))

            val keys = jwksMap["keys"] as List<Map<String, Any>>?

            // Find the key with the matching key ID (kid)
            for (key in keys!!) {
                val kid = key["kid"] as String?
                if (kid != null && kid == keyId) {
                    // Convert the key to a PublicKey (e.g., RSA)
                    val alg = key["alg"] as String?
                    val e = key["e"] as String?
                    val n = key["n"] as String?
                    val eBytes = Base64.getUrlDecoder().decode(e)
                    val nBytes = Base64.getUrlDecoder().decode(n)
                    val modulus = BigInteger(1, nBytes)
                    val exponent = BigInteger(1, eBytes)

                    // Construct RSA public key
                    val keySpec = RSAPublicKeySpec(modulus, exponent)
                    val keyFactory = KeyFactory.getInstance("RSA")
                    return keyFactory.generatePublic(keySpec)
                }
            }
            return null
        }
    }
}